### PR TITLE
[MM-60083] export: do not export DMs/GMs if users are deleted

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
A separate PR will accommodate this one to convert DMs/GMs to private channel if any of the users are deleted. For now, to keep things simple and straight we should not export incomplete channels (structurally). 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-60083

#### Release Note

```release-note
If any of the users of a DM or GM channel is permanently deleted, that channel will not be exported for Bulk Export workflow.
```

### CODE REVIEW

File: server/channels/app/export.go [hunk 1]
```diff
@@ -125,6 +125,12 @@ func (a *App) BulkExport(ctx request.CTX, writer io.Writer, outPath string, job
 		job.Data = make(model.StringMap)
 	}
 
+	if job == nil {
+		job = &model.Job{
+			Data: make(model.StringMap),
+		}
+	}
```
Comment (streamer45) [inline server/channels/app/export.go:132]: A couple of questions:

1. When would the job be `nil`?
2. Should we perhaps consolidate to one call only by reworking the above check? (lines 124-125).

File: server/channels/app/export.go [hunk 2]
```diff
@@ -878,14 +884,38 @@ func (a *App) exportAllDirectChannels(ctx request.CTX, job *model.Job, writer io
 		for _, channel := range channels {
 			afterId = channel.Id
 
+			// Skip deleted.
+			if channel.DeleteAt != 0 {
+				continue
+			}
```
Comment (streamer45) [inline server/channels/app/export.go:890]: Not sure about this. Exporting archived channels makes sense to me. They are even visible, depending on `ExperimentalViewArchivedChannels` setting.

Comment (isacikgoz) [reply server/channels/app/export.go:890]: It was there before, funnily enough you can't actually delete GMs or DMs 🤷 So I though it may have some legacy reasons to keep it there.

**Review by streamer45 — APPROVED** [review]

File: server/channels/app/export.go [hunk 3]
```diff
@@ -121,7 +121,11 @@ func (a *App) BulkExport(ctx request.CTX, writer io.Writer, outPath string, job
 		}
 	}
 
-	if job != nil && job.Data == nil {
+	if job == nil {
```
Comment (agnivade) [inline server/channels/app/export.go:124]: Is this related to the PR?

File: server/channels/app/export.go (hunk 2 above)
Comment (agnivade) [reply server/channels/app/export.go:888]: I believe this is wrong. Because `includeArchivedChannels` is a param passed to the function which is respected in the SQL query. So the channels contain archived channels depending on that flag, therefore there shouldn't be a reason to check for `DeleteAt` again. But it's unrelated to this PR, I agree.

File: server/channels/app/export.go [hunk 4]
```diff
@@ -878,14 +882,38 @@ func (a *App) exportAllDirectChannels(ctx request.CTX, job *model.Job, writer io
...
+			case model.ChannelTypeGroup:
+				groupMembers := make([]string, len(channel.Members))
+				for i, m := range channel.Members {
+					groupMembers[i] = m.UserId
+				}
+				if channel.Name != model.GetGroupNameFromUserIds(groupMembers) {
+					// this is the case of a group channel when other user is permanently deleted
+					// we skip this channel and inform by logging it.
+					job.Data["skipped_direct_channels"] = job.Data["skipped_direct_channels"] + "," + channel.Id
+					ctx.Logger().Info("Skipping unintegrated group channel", mlog.String("channel_id", channel.Id))
```
Comment (agnivade) [inline server/channels/app/export.go:906]: "unintegrated" sounds like an odd choice of word. I'd rather say "Skipping GM channel with partially deleted members", and also log it an WARN level while we're at it, because it's a change in behavior.

**Review by agnivade — COMMENTED** [review]
Some minor questions. Looking good.

File: server/channels/app/export.go (hunk 3 above)
Comment (isacikgoz) [reply server/channels/app/export.go:124]: So we now rely on `job` being not `nil` and it's generally passed as `nil` in the tests, so that's why we are creating it.

**Review by agnivade — APPROVED** [review]
